packages/financial-templates-lib/src/price-feed/UniswapPriceFeed.js

```diff
@@ -55,16 +77,20 @@ class UniswapPriceFeed extends PriceFeedInterface {
 
   // Not part of the price feed interface. Can be used to pull the uniswap price at the most recent block.
   getLastBlockPrice() {
-    return this.lastBlockPrice;
+    return this.lastBlockPrice && this.convertDecimals(this.lastBlockPrice);
   }
 
   getPriceFeedDecimals() {
-    return 18;
+    return this.decimals;
   }
 
   async update() {
-    // TODO: optimize this call. This may be very slow or break if there are many transactions.
-    const events = await this.uniswap.getPastEvents("Sync", { fromBlock: 0 });
+    // Approximate the first block from which we'll need price data from based on the
+    // lookback and twap length:
+    const lookbackWindow = this.twapLength + this.historicalLookback;
+    const latestBlockNumber = (await this.web3.eth.getBlock("latest")).number;
+    const earliestBlockNumber = latestBlockNumber - Math.ceil(lookbackWindow / (await averageBlockTimeSeconds()));
```

Comment:
1. Should we cushion this a little so we don't undershoot in case the average is off?
2. What happens if the pool just doesn't trade much? (i.e. the price is constant) for the entire lookback period?

Reply:
1. I wanted to avoid hardcoding a value but I'll add a cushion of 10%
2. This is a really good point, perhaps in this case where there are no events in the lookback then we should crawl backwards until we find a Sync event.

Reply:
1. Implemented a hard-coded cushion of 10% for 1, (and changed the Balancer price feed's lookback cushion to be a % rather than a fixed # of blocks)
2. Added a while loop to crawl backwards for events until either `events.length > 0` or the `fromBlock` passed to `getPastEvents` is 0